File: python/rd_api_sdk/reality_data.py
from typing import Optional, List
from enum import Enum
from datetime import datetime
from dateutil import parser


class Classification(str, Enum):
    """
    Classification of a reality data
    """
    TERRAIN = "Terrain"
    IMAGERY = "Imagery"
    PINNED = "Pinned"
    MODEL = "Model"
    UNDEFINED = "Undefined"
 
    @staticmethod
    def from_str(s: str):
        for x in Classification:
            if x == x.value:
                return x


class Visibility(str, Enum):
    """
    Visibility of a reality data
    """
    PUBLIC = "PUBLIC"
    ENTERPRISE = "ENTERPRISE"
    PERMISSION = "PERMISSION"
    PRIVATE = "PRIVATE"

    @staticmethod
    def from_str(s: str):
        for x in Visibility:
            if x == x.value:
                return x


class Coordinate:
    """
    Coordinate in latitude/longitude format
    """
    def __init__(self, lat: float, long: float):
        """
        Constructor

        :param lat: Latitude in numeric format (40.0654)
        :param long: Longitude in numeric format (-75.6877)
        """
        self._lat = lat
        self._long = long

    def lat(self) -> float:
        """
        :return: Latitude
        """
        return self._lat

    def long(self) -> float:
        """
        :return: Longitude
        """
        return self._long


class Extent:
    """
    Extent of a reality data
    """
    def __init__(self, southwest: Coordinate, northeast: Coordinate):
        self._sw = southwest
        self._ne = northeast

    def southwest(self, c: Optional[Coordinate] = None) -> Coordinate:
        """
        Southwest corner of the extent

        :param c: Optional Coordinate to replace the current value
        :return: Southwest coordinate of the extent
        """
        if c is not None:
            self._sw = c
        return self._sw

    def northeast(self, c: Optional[Coordinate] = None) -> Coordinate:
        """
        Northeast corner of the extent

        :param c: Optional Coordinate to replace the current value
        :return: Northeast coordinate of the extent
        """
        if c is not None:
            self._ne = c
        return self._ne


class Acquisition:
    """
    Details about data acquisition
    """

    def __init__(self, start_date_time: Optional[datetime] = None,
                 end_date_time: Optional[datetime] = None, acquirer: Optional[str] = None):
        """
        Constructor

        :param start_date_time: Optional acquisition start date time
        :param end_date_time: Optional acquisition end date time
        :param acquirer: Optional acquirer name
        """
        self._start_date = start_date_time
        self._end_date = end_date_time
        self._acquirer = acquirer

    def start_date(self) -> Optional[datetime]:
        """
        :return: Optional start date of the acquisition
        """
        return self._start_date

    def end_date(self) -> Optional[datetime]:
        """
        :return: Optional end date of the acquisition
        """
        return self._end_date

    def acquirer(self) -> Optional[str]:
        """
        :return: Optional acquirer
        """
        return self._acquirer


class RealityDataUpdate:
    def __init__(self, name: Optional[str] = None, classification: Optional[Classification] = None,
                 rd_type: Optional[str] = None, acquisition: Optional[Acquisition] = None,
                 description: Optional[str] = None, dataset: Optional[str] = None,
                 group: Optional[str] = None, root_doc: Optional[str] = None, extent: Optional[Extent] = None,
                 authoring: Optional[bool] = None):
        """
        Constructor

        :param name: Optional new reality data name
        :param classification: Optional new classification
        :param rd_type: Optional new reality data type
        :param acquisition: Optional new acquisition information
        :param description: Optional new description
        :param dataset: Optional new loose dataset grouping
        :param group: Optional new loose second level of grouping
        :param root_doc: Optional new root document of the reality data
        :param extent: Optional new extent of the reality data
        :param authoring: Optional new authoring status of the reality data
        """
        self._name = name
        self._classification = classification
        self._type = rd_type
        self._acquisition = acquisition
        self._description = description
        self._dataset = dataset
        self._group = group
        self._root_doc = root_doc
        self._extent = extent
        self._authoring = authoring

    def name(self) -> Optional[str]:
        return self._name

    def classification(self) -> Optional[Classification]:
        return self._classification

    def type(self) -> Optional[str]:
        return self._type

    def acquisition(self) -> Optional[Acquisition]:
        return self._acquisition

    def description(self) -> Optional[str]:
        return self._description

    def dataset(self) -> Optional[str]:
        return self._dataset

    def group(self) -> Optional[str]:
        return self._group

    def root_document(self) -> Optional[str]:
        return self._root_doc

    def extent(self) -> Optional[Extent]:
        return self._extent

    def authoring(self) -> Optional[bool]:
        return self._authoring


class RealityDataCreate(RealityDataUpdate):
    def __init__(self, name: str, classification: Classification, rd_type: str,
                 acquisition: Optional[Acquisition] = None,
                 description: Optional[str] = None, dataset: Optional[str] = None,
                 group: Optional[str] = None, root_doc: Optional[str] = None, extent: Optional[Extent] = None,
                 authoring: Optional[bool] = None):
        """
        Constructor

        :param name: Reality data name
        :param classification: Reality data classification
        :param rd_type: Reality data type
        :param acquisition: Reality data acquisition information (type is mandatory)
        :param description: Optional description
        :param dataset: Optional loose dataset grouping
        :param group: Optional loose second level of grouping
        :param root_doc: Optional root document of the reality data
        :param extent: Optional extent of the reality data
        :param authoring: Optional authoring status of the reality data
        """
        RealityDataUpdate.__init__(self, name, classification, rd_type, acquisition, description, dataset, group,
                                   root_doc, extent, authoring)

    def name(self) -> str:
        return self._name

    def classification(self) -> Classification:
        return self._classification

    def type(self) -> str:
        return self._type


class RealityData(RealityDataCreate):
    def __init__(self, rd_id: str, name: str, classification: Classification, rd_type: str,
                 modified_date_time_str: str, last_access_date_time_str: str, created_date_time_str: str,
                 datacenter: str,
                 acquisition: Optional[Acquisition],
                 description: Optional[str] = None, dataset: Optional[str] = None,
                 group: Optional[str] = None, root_doc: Optional[str] = None,
                 extent: Optional[Extent] = None, authoring: Optional[bool] = None):
        """
        Constructor

        :param rd_id: Reality data id
        :param name: Reality data name
        :param classification: Reality data classification
        :param rd_type: Reality data type
        :param modified_date_time_str: Last modified datetime as string
        :param last_access_date_time_str:  Last access datetime as string
        :param created_date_time_str: Created datetime as string
        :param datacenter: Datacenter location of the reality data
        :param acquisition: Reality data acquisition information
        :param description: Optional description
        :param dataset: Optional loose dataset grouping
        :param group: Optional loose second level of grouping
        :param root_doc: Optional root document of the reality data
        :param extent: Optional extent of the reality data
        :param authoring: Optional authoring status of the reality data
        """
        RealityDataCreate.__init__(self, name, classification, rd_type, acquisition, description, dataset, group,
                                   root_doc, extent, authoring)

        self._id = rd_id
        self._modified = parser.parse(modified_date_time_str)
        self._last_access = parser.parse(last_access_date_time_str)
        self._created = parser.parse(created_date_time_str)
        self._datacenter = datacenter

    def id(self) -> str:
        return self._id

    def modified(self) -> datetime:
        return self._modified

    def last_access(self) -> datetime:
        return self._last_access

    def created(self) -> datetime:
        return self._created

    def datacenter_location(self) -> str:
        return self._datacenter

File: typescript/packages/reality-data-client/src/RealityData.ts
/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { RealityDataClientError } from "./RealityDataClientError";
import type { RealityDataAccessClient } from "./RealityDataClient";
import { getRequestConfig } from "./RequestOptions";

/**
 * Extent of a reality data, delimited by southwest and northeast coordinates.
 */
export interface Extent {
  southWest: Point;
  northEast: Point;
}

/**
 * Point used to define an extent.
 */
export interface Point {
  latitude: number;
  longitude: number;
}

/**
 * Coordinate Reference System (CRS) used for the reality data.
 * It is restricted to specific reality data types, read documentation here:
 * https://developer.bentley.com/apis/reality-management/rm-rd-details/#crs
 */
export interface Crs {
  id: string;
  verticalId?: string;
}

/**
 * Provides information regarding the acquisition, such as dates and acquirer used.
 */
export interface Acquisition {
  startDateTime: Date;
  endDateTime?: Date;
  acquirer?: string;
}

/**
 * Cache parameters for reality data access. Contains the blob url, the timestamp to refresh (every 50 minutes) the url and the root document path.
 * Cache contains one value for the read permission url, and one of the write permission.
 * */
class ContainerCache {

  private _containerRead?: ContainerCacheValue;
  private _containerWrite?: ContainerCacheValue;

  public getCache(access: string): ContainerCacheValue | undefined {
    if (access === "Read")
      return this._containerRead;
    else
      return this._containerWrite;
  }

  public setCache(containerCacheValue: ContainerCacheValue, access: string) {
    if (access === "Read")
      this._containerRead = containerCacheValue;
    else
      this._containerWrite = containerCacheValue;
  }
}

interface ContainerCacheValue {
  url: URL;
  timeStamp: Date;
}

/** RealityData
 * This class implements a Reality Data instance.
 * Data is accessed directly through methods of the reality data instance.
 * Access to the data required a properly entitled token though the access to the blob is controlled through
 * an Azure blob URL, the token may be required to obtain this Azure blob URL or refresh it.
 * The Azure blob URL is considered valid for an hour and is refreshed after 50 minutes.
 * In addition to the reality data properties, and Azure blob URL and internal states, a reality data also contains
 * the identification of the iTwin to be used for access permissions and
 * may contain a RealityDataClient to obtain the specialization to communicate with Reality Management API (to obtain the Azure blob URL).
 * @beta
 */
export class ITwinRealityData {

  public id: string;
  public displayName?: string;
  public dataset?: string;
  public group?: string;
  public attribution?: string;
  public termsOfUse?: string;
  public dataCenterLocation?: string;
  public description?: string;
  public rootDocument?: string;
  public tags?: string[];
  public acquisition?: Acquisition;
  public size?: number;
  public authoring?: boolean;
  public classification?: string;
  public type?: string;
  public extent?: Extent;
  public crs?: Crs;
  /** @deprecated in 1.0.1 not used in Reality Management API. Will be removed in next major update.*/
  public accessControl?: string; // TODO: remove in next major update
  public modifiedDateTime?: Date;
  public lastAccessedDateTime?: Date;
  public createdDateTime?: Date;
  public ownerId?: string;

  /** Link to client to fetch the blob url */
  public client: undefined | RealityDataAccessClient;

  /** The GUID identifier of the iTwin used when using the client. */
  public iTwinId: string;

  /** Cache parameters for reality data access. Contains the blob url, the timestamp to refresh (every 50 minutes) the url and the root document path. */
  private _containerCache: ContainerCache;

  /**
   * Creates an instance of RealityData.
   * @beta
   */
  public constructor(client: RealityDataAccessClient, realityData?: any, iTwinId?: string) {

    this.client = client;
    this._containerCache = new ContainerCache();

    if (realityData) {
      this.id = realityData.id;
      this.displayName = realityData.displayName;
      this.dataset = realityData.dataset;
      this.group = realityData.group;
      this.attribution = realityData.attribution;
      this.termsOfUse = realityData.termsOfUse;
      this.dataCenterLocation = realityData.dataCenterLocation;
      this.description = realityData.description;
      this.rootDocument = realityData.rootDocument;
      this.tags = realityData.tags;
      if (realityData.acquisition) {
        this.acquisition = (realityData.acquisition as Acquisition);
        this.acquisition.startDateTime = new Date(realityData.acquisition.startDateTime);
        this.acquisition.endDateTime = realityData.acquisition.endDateTime ? new Date(realityData.acquisition.endDateTime) : undefined;
        this.acquisition.acquirer = realityData.acquisition.acquirer ? realityData.acquisition.acquirer : undefined;
      }
      this.size = realityData.size;
      this.authoring = realityData.authoring;
      this.classification = realityData.classification;
      this.type = realityData.type;
      this.extent = realityData.extent;
      this.crs = realityData.crs;
      this.accessControl = realityData.accessControl;
      this.modifiedDateTime = new Date(realityData.modifiedDateTime);
      this.lastAccessedDateTime = new Date(realityData.lastAccessedDateTime);
      this.createdDateTime = new Date(realityData.createdDateTime);
      this.ownerId = realityData.ownerId;
    }

    if (iTwinId)
      this.iTwinId = iTwinId;
  }

  /**
     * Gets string url to fetch blob data from. Access is read-only.
     * @param accessToken The client request context.
     * @param blobPath name or path of tile
     * @param writeAccess Optional boolean indicating if write access is requested. Default is false for read-only access. The realitydata:modify scope is required to grant the "write" access.
     * @returns string url for blob data
     * @beta
     */
  public async getBlobUrl(accessToken: string, blobPath: string, writeAccess = false): Promise<URL> {
    const accessTokenResolved = await this.resolveAccessToken(accessToken);
    const url = await this.getContainerUrl(accessTokenResolved, writeAccess);
    if (blobPath === undefined)
      return url;

    const host = `${url.origin + url.pathname}/`;
    const query = url.search;
    return new URL(`${host}${blobPath}${query}`);
  }
  /**
   * Try to use authorizationClient in RealityDataClientOptions to get the access token
   * otherwise, will return the input token
   * This is a workaround to support different authorization client for the reality data client and iTwin-core.
   */
  private async resolveAccessToken(accessToken: string): Promise<string> {
    return this.client?.authorizationClient ? this.client.authorizationClient.getAccessToken() : accessToken;
  }

  /**
     * Gets a tile access url URL object
     * @param accessToken The client request context.
     * @param writeAccess Optional boolean indicating if write access is requested. Default is false for read-only access.
     * @returns app URL object for blob url
     * @beta
     */
  private async getContainerUrl(accessToken: string, writeAccess = false): Promise<URL> {

    if (!this.client)
      throw new RealityDataClientError(422, "Invalid container request (RealityDataAccessClient is not set).");

    const access = (writeAccess === true ? "Write" : "Read");
    const accessTokenResolved = await this.resolveAccessToken(accessToken);

    try {

      const containerCache = this._containerCache.getCache(access);
      const blobUrlRequiresRefresh = !containerCache?.timeStamp || (Date.now() - containerCache?.timeStamp.getTime()) > 3000000; // 3 million milliseconds or 50 minutes

      if (undefined === containerCache?.url || blobUrlRequiresRefresh) {

        const url = new URL(`${this.client.baseUrl}/${this.id}/${ writeAccess === true ? "writeAccess" : "readAccess"}`);

        if(this.iTwinId)
          url.searchParams.append("iTwinId", this.iTwinId);

        const requestOptions = getRequestConfig(accessTokenResolved, "GET", this.client.apiVersion);

        const response = await fetch(url.href, requestOptions);

        if (!response.ok) {
          throw new RealityDataClientError(response.status, "Invalid container request.");
        }

        const responseData = await response.json();

        if (!responseData?._links?.containerUrl?.href) {
          throw new RealityDataClientError(422, "Invalid container request (API returned an unexpected response).");
        }

        // update cache
        const newContainerCacheValue: ContainerCacheValue = {
          url: new URL(responseData._links.containerUrl.href),
          timeStamp: new Date(Date.now()),
        };

        this._containerCache.setCache(newContainerCacheValue, access);
      }
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return this._containerCache.getCache(access)!.url;
    } catch {
      throw new RealityDataClientError(422, "Invalid container request.");
    }
  }
}